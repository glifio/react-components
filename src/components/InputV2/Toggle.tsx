import styled, { css } from 'styled-components'
import PropTypes from 'prop-types'
import { Label } from './Label'

const ToggleLabel = styled(Label)`
  .toggle-wrapper {
    position: relative;
    padding: 0.5em 0;
  }

  input {
    position: absolute;
    top: 50%;
    left: 50%;
    width: 0;
    height: 0;
    opacity: 0;
  }

  .toggle {
    display: inline-block;
    position: relative;
    width: 2.25em;
    height: 1.5em;
    border-radius: 0.75em;
    border: 2px solid var(--blue-light);
    background-color: var(--blue-light);

    ${props =>
      props.disabled &&
      css`
        border-color: var(--gray-light) !important;
        background-color: var(--gray-light) !important;
      `}

    &::after {
      content: '';
      position: absolute;
      top: 0;
      left: 0;
      width: calc(1.5em - 4px);
      height: calc(1.5em - 4px);
      border-radius: 50%;
      background-color: var(--white);
    }
  }

  input:hover ~ .toggle,
  input:checked ~ .toggle,
  input:focus ~ .toggle {
    border-color: var(--blue-medium);
    background-color: var(--blue-medium);
  }

  input:checked ~ .toggle::after {
    left: auto;
    right: 0;
  }
`

export const Toggle = ({
  label,
  autofocus,
  disabled,
  controlled,
  checked,
  onChange
}: ToggleProps) => (
  <ToggleLabel disabled={disabled}>
    <span>{label}</span>
    <span className='toggle-wrapper'>
      <input
        type='checkbox'
        autoFocus={autofocus}
        disabled={disabled}
        checked={controlled ? checked : undefined}
        defaultChecked={controlled ? undefined : checked}
        onChange={e => onChange(e.target.checked)}
      />
      <span className='toggle'></span>
    </span>
  </ToggleLabel>
)

interface ToggleProps {
  label: string
  autofocus: boolean
  disabled: boolean
  controlled: boolean
  checked: boolean
  onChange: (checked: boolean) => void
}

Toggle.propTypes = {
  label: PropTypes.string.isRequired,
  autofocus: PropTypes.bool,
  disabled: PropTypes.bool,
  controlled: PropTypes.bool,
  checked: PropTypes.bool,
  onChange: PropTypes.func
}

Toggle.defaultProps = {
  autofocus: false,
  disabled: false,
  controlled: true,
  checked: false,
  onChange: () => {}
}
