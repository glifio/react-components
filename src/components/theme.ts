export const baseColors = {
  mono: {
    black: '#000',
    nearblack: '#262626',
    darkgray: '#666666',
    lightgray: '#C4C4C4',
    silver: '#999999',
    white: '#ffffff',
    transparent: 'transparent'
  },
  blue: {
    lightest: '#F6F8FE',
    lighter: '#EFF3FD',
    light: '#E4EBFC',
    mid: '#d1ddfa',
    mid2: '#bacbf7',
    muted: '#e8ecf7',
    muted700: '#d9e0f2'
  },
  green: {
    primary: '#1AD08F',
    light: '#D2F5ED',
    muted: '#85e0cb',
    dark: '#007056',
    darker: '#08442F'
  },
  red: {
    light: '#FC6D6F',
    dark: '#660000',
    base: '#ff4242'
  },
  orange: {
    light: '#FF9E80',
    dark: '#661800',
    mid: '#FFA826'
  },
  yellow: {
    light: '#FFDC99',
    deep: '#FCA703',
    mid: '#F7C103'
  },
  purple: {
    light: '#E0D7FE',
    mid: '#5E26FF',
    deep: '#290a85'
  }
}

// The core color object is the only non-semantic object here. This is because these colors are used so widely that it would be highly redundant to replicate these color values repeatedly within this "colors" object to style the text/background of every single Component the app renders.
const core = {
  primary: baseColors.purple.mid,
  secondary: baseColors.purple.light,
  tertiary: baseColors.purple.deep,
  black: baseColors.mono.black,
  nearblack: baseColors.mono.nearblack,
  darkgray: baseColors.mono.darkgray,
  lightgray: baseColors.mono.lightgray,
  silver: baseColors.mono.silver,
  white: baseColors.mono.white,
  transparent: baseColors.mono.transparent,
  yellow: baseColors.yellow.deep
}

const colors = {
  core,
  buttons: {
    primary: {
      background: baseColors.green.primary,
      borderColor: baseColors.green.primary,
      color: baseColors.green.darker
    },
    secondary: {
      background: baseColors.mono.transparent,
      borderColor: baseColors.mono.nearblack,
      color: baseColors.mono.nearblack
    },
    tertiary: {
      background: baseColors.mono.transparent,
      borderColor: baseColors.purple.light,
      color: baseColors.purple.light
    },
    'tertiary-selected': {
      background: baseColors.purple.mid,
      borderColor: baseColors.purple.mid,
      color: baseColors.purple.light
    }
  },
  background: {
    app: baseColors.blue.lightest,
    screen: baseColors.blue.lighter,
    text: baseColors.blue.light,
    messageHistory: baseColors.blue.muted
  },
  card: {
    account: {
      background: core.primary,
      color: core.secondary
    },
    balance: {
      background: baseColors.mono.transparent,
      color: core.nearblack
    },
    confirmation: {
      background: baseColors.green.light,
      foreground: baseColors.green.dark
    },
    error: {
      background: baseColors.red.light,
      foreground: baseColors.red.dark
    }
  },
  input: {
    background: {
      base: baseColors.blue.mid,
      active: baseColors.blue.mid2,
      valid: baseColors.green.muted,
      invalid: baseColors.red.light
    },
    border: core.silver
  },
  status: {
    success: {
      background: baseColors.green.primary,
      foreground: baseColors.green.darker
    },
    pending: {
      background: baseColors.yellow.deep,
      foreground: baseColors.mono.darkgray
    },
    warning: {
      background: baseColors.orange.light,
      foreground: baseColors.orange.dark
    },
    fail: {
      background: baseColors.red.light,
      foreground: baseColors.red.dark
    },
    inactive: baseColors.mono.lightgray
  },
  // gonna start slowly migrating colors from their inheritance vars above
  blue: {
    muted700: '#d9e0f2'
  },
  gray: {
    light: '#C4C4C4',
    medium: '#949598',
    dark: '#666666'
  },
  green: {
    primary: '#1AD08F',
    light: '#D2F5ED',
    muted: '#85e0cb',
    medium: '#007B39',
    dark: '#007056',
    darker: '#08442F'
  },
  yellow: {
    light: '#FFDC99',
    medium: '#C59900',
    deep: '#FCA703',
    mid: '#F7C103'
  },
  purple: {
    light: '#E0D7FE',
    medium: '#5E26FF',
    deep: '#290a85'
  }
}

const theme = {
  colors,
  fontSizes: [
    '0rem',
    '1rem',
    '1.125rem',
    '1.25rem',
    '1.5rem',
    '2rem',
    '3rem',
    '5rem'
  ],
  fontWeights: [0, 400, 600, 900],
  letterSpacings: [0, 1, 2, 4, 8],
  lineHeights: {
    solid: 1,
    title: 1.2,
    copy: 1.4
  },
  textStyles: {
    header: {
      fontSize: 'calc(48px + (48 - 32) * (100vw - 360px) / (1440 - 360))',
      fontWeight: 400,
      margin: 0,
      lineHeight: 'title',
      fontFamily: 'RT-Alias-Grotesk'
    },
    bigTitle: {
      fontSize: 6,
      fontWeight: 700,
      margin: 0,
      lineHeight: 'solid',
      fontFamily: 'RT-Alias-Medium'
    },
    title: {
      fontSize: 4,
      fontWeight: 400,
      margin: 0,
      lineHeight: 'title',
      fontFamily: 'RT-Alias-Grotesk'
    },
    text: {
      fontSize: 2,
      fontColor: core.nearblack,
      fontWeight: 400,
      lineHeight: 'copy',
      fontFamily: 'RT-Alias-Grotesk'
    },
    label: {
      fontSize: 2,
      fontColor: core.darkgray,
      fontWeight: 400,
      lineHeight: 'solid',
      fontFamily: 'RT-Alias-Grotesk'
    },
    num: {
      m: {
        fontSize: 4,
        fontWeight: 400,
        margin: 0,
        lineHeight: 'title',
        fontFamily: 'RT-Alias-Grotesk'
      },
      l: {
        fontSize: 5,
        fontWeight: 700,
        margin: 0,
        lineHeight: 'title',
        fontFamily: 'RT-Alias-Grotesk'
      },
      xl: {
        fontSize: 6,
        fontWeight: 700,
        margin: 0,
        lineHeight: 'solid',
        fontFamily: 'RT-Alias-Grotesk'
      },
      xxl: {
        fontSize: 7,
        fontWeight: 700,
        margin: 0,
        lineHeight: 'solid',
        fontFamily: 'RT-Alias-Grotesk'
      }
    }
  },
  fonts: {
    AliasMedium:
      '"RT-Alias-Medium", "system-ui", "Segoe UI", "Roboto", Helvetica',
    AliasGrotesk:
      '"RT-Alias-Grotesk", "system-ui", "Segoe UI", "Roboto", Helvetica',
    sansSerif:
      '"system-ui", "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";',
    mono: '"system-ui", "Segoe UI", Roboto Mono, Helvetica, Arial, monospace, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";'
  },
  space: [0, 4, 8, 16, 24, 32, 48, 64, 128, 256],
  sizes: [
    0, 4, 8, 16, 24, 32, 48, 64, 80, 160, 240, 300, 480, 560, 640, 720, 800,
    960, 1024, 1200, 1440
  ],
  radii: ['0', '1px', '4px', '8px', '16px', '32px', '100px'],
  borders: [0, `1px solid`, `4px solid`],
  borderStyles: ['solid'],
  borderWidths: ['0', '1px', '2px', '4px'],
  breakpoints: ['40em', '52em', '64em'],
  shadows: [
    '0',
    '0 0.7px 2.2px -8px rgba(0, 0, 0, 0.18),0 1.7px 5.4px  rgba(0, 0, 0, 0.129),0 3.1px 10.1px rgba(0, 0, 0, 0.107),0 5.6px 18.1px rgba(0, 0, 0, 0.09),0 10.4px 33.8px rgba(0, 0, 0, 0.073),0 25px 81px rgba(0, 0, 0, 0.051)',
    'rgba(0, 0, 0, 0.10) 0px 0.7px 2.2px -8px, rgba(0, 0, 0, 0.04) 0px 1.7px 2.4px, rgba(0, 0, 0, 0.106) 0px 3.1px 8.1px, rgba(0, 0, 0, 0.04) 0px 5.6px 12.1px, rgba(0, 0, 0, 0.045) 0px 4.4px 4.8px, rgba(0, 0, 0, 0.05) 0px 15px 41px'
  ],
  opacity: {
    disabled: 0.4
  },
  zIndices: [0, 9, 99, 999, 9999]
}

export default theme

/* EVERYTHING BELOW HERE IS THEMEV2 */

// safe to change
const deviceWidthPxs = {
  phone: 480,
  tablet: 868,
  desktop: 1124
}

const spaces = {
  phone: {
    sm: '',
    small: '',
    default: '8px',
    md: '8px',
    medium: '8px',
    lg: '20px',
    large: '20px'
  },
  default: {
    sm: '',
    small: '',
    default: '10px',
    md: '10px',
    medium: '10px',
    lg: '20px',
    large: '20px'
  },
  tablet: {
    sm: '',
    small: '',
    default: '10px',
    md: '10px',
    medium: '10px',
    lg: '20px',
    large: '20px'
  },
  desktop: {
    sm: '',
    small: '',
    default: '10px',
    md: '10px',
    medium: '10px',
    lg: '20px',
    large: '20px'
  }
}

const fontSizes = {
  phone: {
    sm: '13px',
    small: '13px',
    default: '16px',
    md: '18px',
    medium: '18px',
    lg: '20px',
    large: '20px'
  },
  default: {
    sm: '13px',
    small: '13px',
    default: '18px',
    md: '26px',
    medium: '26px',
    lg: '28px',
    large: '28px'
  },
  tablet: {
    sm: '13px',
    small: '13px',
    default: '18px',
    md: '26px',
    medium: '26px',
    lg: '28px',
    large: '28px'
  },
  desktop: {
    sm: '13px',
    small: '13px',
    default: '18px',
    md: '26px',
    medium: '26px',
    lg: '28px',
    large: '28px'
  }
}

export const maxWidth = '1450px'

// dont change
export const devices = {
  phone: `${deviceWidthPxs.phone}px`,
  tablet: `${deviceWidthPxs.tablet}px`,
  desktop: `${deviceWidthPxs.desktop}px`
}

// these are
// one of size = sm, default, md, lg
// one of device = phone, tablet, desktop
export function space(size = 'default', device = 'default') {
  return spaces[device][size]
}

export function fontSize(size = 'default', device = 'default') {
  return fontSizes[device][size]
}
