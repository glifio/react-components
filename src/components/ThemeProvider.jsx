import PropTypes from 'prop-types'
import { ThemeProvider as StyledThemeProvider } from 'styled-components'
import { theme as defaultTheme } from './theme'

const ThemeProvider = ({ children, theme, ...rest }) => {
  return (
    <StyledThemeProvider theme={theme} {...rest}>
      {children}
    </StyledThemeProvider>
  )
}

ThemeProvider.propTypes = {
  theme: PropTypes.object,
  children: PropTypes.node
}

ThemeProvider.defaultProps = {
  theme: defaultTheme,
  children: <></>
}

ThemeProvider.displayName = 'ThemeProvider'

export default ThemeProvider
