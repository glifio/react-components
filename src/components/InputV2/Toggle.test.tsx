import { cleanup, render, act, RenderResult } from '@testing-library/react'
import { Toggle } from './Toggle'
import ThemeProvider from '../ThemeProvider'
import theme from '../theme'

const labelText = 'Turn me on or off'
const infoText = 'And see what it does'

describe('Toggle input', () => {
  afterEach(cleanup)

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('it renders the on state correctly', async () => {
    let result: RenderResult | null = null
    await act(async () => {
      result = render(
        <ThemeProvider theme={theme}>
          <Toggle
            label={labelText}
            info={infoText}
            checked={true}
          />
        </ThemeProvider>
      )
    })
    expect(result.container.firstChild).toMatchSnapshot()
  })

  test('it renders the off state correctly', async () => {
    let result: RenderResult | null = null
    await act(async () => {
      result = render(
        <ThemeProvider theme={theme}>
          <Toggle
            label={labelText}
            info={infoText}
            checked={false}
          />
        </ThemeProvider>
      )
    })
    expect(result.container.firstChild).toMatchSnapshot()
  })
})
