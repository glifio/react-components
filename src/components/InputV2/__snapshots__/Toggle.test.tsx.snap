// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle input it renders the off state correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  text-align: left;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 > *:first-child {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 > *:last-child:not(.toggle-wrapper) {
  -webkit-flex: 0 0 50%;
  -ms-flex: 0 0 50%;
  flex: 0 0 50%;
}

.c0 span.info,
.c0 span.error {
  font-size: 0.875rem;
}

.c0 span.error {
  color: var(--red-medium);
}

.c0 .button-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
}

.c0 .button-wrapper > *:first-child {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

.c0 .button-wrapper > *:not(:first-child) {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-transition: -webkit-transform 0.1s ease-out;
  -webkit-transition: transform 0.1s ease-out;
  transition: transform 0.1s ease-out;
}

.c0 .button-wrapper > *:not(:first-child):hover:not(:active) {
  -webkit-transform: scale(1.2);
  -ms-transform: scale(1.2);
  transform: scale(1.2);
}

.c0 .button-wrapper > *:not(:first-child):active {
  -webkit-transition: none;
  transition: none;
}

.c0 .unit-wrapper {
  position: relative;
}

.c0 .unit-wrapper input {
  width: 100%;
}

.c0 .unit-wrapper .unit {
  position: absolute;
  top: 50%;
  right: 1em;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  color: var(--purple-medium);
}

.c0 .select-wrapper {
  position: relative;
}

.c0 .select-wrapper select {
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c0 .select-wrapper .select-arrow {
  position: absolute;
  pointer-events: none;
  top: 50%;
  right: 1.2em;
  width: 0.6em;
  height: 0.6em;
  border-right: 1.5px solid var(--black);
  border-bottom: 1.5px solid var(--black);
  -webkit-transform: translateY(-0.35em) rotate(45deg);
  -ms-transform: translateY(-0.35em) rotate(45deg);
  transform: translateY(-0.35em) rotate(45deg);
}

.c1 .toggle-wrapper {
  position: relative;
}

.c1 input {
  position: absolute;
  top: 50%;
  left: 50%;
  width: 0;
  height: 0;
  opacity: 0;
}

.c1 .toggle {
  display: inline-block;
  position: relative;
  width: 2.25em;
  height: 1.5em;
  border-radius: 0.75em;
  border: 2px solid var(--blue-medium);
  background-color: var(--blue-medium);
}

.c1 .toggle::after {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: calc(1.5em - 4px);
  height: calc(1.5em - 4px);
  border-radius: 50%;
  background-color: var(--white);
}

.c1 input:checked ~ .toggle {
  border-color: var(--blue-dark);
  background-color: var(--blue-dark);
}

.c1 input:checked ~ .toggle::after {
  left: auto;
  right: 0;
}

<label
  class="c0 c1"
>
  <div>
    <span>
      Turn me on or off
    </span>
    <span
      class="info"
    >
      And see what it does
    </span>
  </div>
  <div
    class="toggle-wrapper"
  >
    <input
      type="checkbox"
    />
    <span
      class="toggle"
    />
  </div>
</label>
`;

exports[`Toggle input it renders the on state correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  text-align: left;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 > *:first-child {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 > *:last-child:not(.toggle-wrapper) {
  -webkit-flex: 0 0 50%;
  -ms-flex: 0 0 50%;
  flex: 0 0 50%;
}

.c0 span.info,
.c0 span.error {
  font-size: 0.875rem;
}

.c0 span.error {
  color: var(--red-medium);
}

.c0 .button-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
}

.c0 .button-wrapper > *:first-child {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

.c0 .button-wrapper > *:not(:first-child) {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-transition: -webkit-transform 0.1s ease-out;
  -webkit-transition: transform 0.1s ease-out;
  transition: transform 0.1s ease-out;
}

.c0 .button-wrapper > *:not(:first-child):hover:not(:active) {
  -webkit-transform: scale(1.2);
  -ms-transform: scale(1.2);
  transform: scale(1.2);
}

.c0 .button-wrapper > *:not(:first-child):active {
  -webkit-transition: none;
  transition: none;
}

.c0 .unit-wrapper {
  position: relative;
}

.c0 .unit-wrapper input {
  width: 100%;
}

.c0 .unit-wrapper .unit {
  position: absolute;
  top: 50%;
  right: 1em;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  color: var(--purple-medium);
}

.c0 .select-wrapper {
  position: relative;
}

.c0 .select-wrapper select {
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c0 .select-wrapper .select-arrow {
  position: absolute;
  pointer-events: none;
  top: 50%;
  right: 1.2em;
  width: 0.6em;
  height: 0.6em;
  border-right: 1.5px solid var(--black);
  border-bottom: 1.5px solid var(--black);
  -webkit-transform: translateY(-0.35em) rotate(45deg);
  -ms-transform: translateY(-0.35em) rotate(45deg);
  transform: translateY(-0.35em) rotate(45deg);
}

.c1 .toggle-wrapper {
  position: relative;
}

.c1 input {
  position: absolute;
  top: 50%;
  left: 50%;
  width: 0;
  height: 0;
  opacity: 0;
}

.c1 .toggle {
  display: inline-block;
  position: relative;
  width: 2.25em;
  height: 1.5em;
  border-radius: 0.75em;
  border: 2px solid var(--blue-medium);
  background-color: var(--blue-medium);
}

.c1 .toggle::after {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: calc(1.5em - 4px);
  height: calc(1.5em - 4px);
  border-radius: 50%;
  background-color: var(--white);
}

.c1 input:checked ~ .toggle {
  border-color: var(--blue-dark);
  background-color: var(--blue-dark);
}

.c1 input:checked ~ .toggle::after {
  left: auto;
  right: 0;
}

<label
  class="c0 c1"
>
  <div>
    <span>
      Turn me on or off
    </span>
    <span
      class="info"
    >
      And see what it does
    </span>
  </div>
  <div
    class="toggle-wrapper"
  >
    <input
      checked=""
      type="checkbox"
    />
    <span
      class="toggle"
    />
  </div>
</label>
`;
